src/graph/executor/query/ShortestPathExecutor.cpp

```diff
@@ -0,0 +1,377 @@
+/* Copyright (c) 2022 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License.
+ */
+#include "graph/executor/query/ShortestPathExecutor.h"
+
+#include "graph/service/GraphFlags.h"
+#include "graph/util/SchemaUtil.h"
+
+using nebula::storage::StorageClient;
+using nebula::storage::StorageRpcResponse;
+using nebula::storage::cpp2::GetNeighborsResponse;
+
+namespace nebula {
+namespace graph {
+
+// ShortestPath
+folly::Future<Status> ShortestPathExecutor::execute() {
+  SCOPED_TIMER(&execTime_);
+
+  auto status = buildRequestDataSet();
+  if (!status.ok()) {
+    return error(std::move(status));
+  }
+  if (cartesian_.empty()) {
+    VLOG(1) << "Empty input.";
+    DataSet emptyResult;
+    return finish(ResultBuilder().value(Value(std::move(emptyResult))).build());
+  }
+
+  for (auto& pair : cartesian_) {
+    // TODO: step 0
+    if (pair.first == pair.second) {
+      return folly::makeFuture<Status>(Status::Error("Source and destination cannot be the same"));
+    }
+    dss_[0].rows = std::vector<Row>{Row(std::vector<Value>{std::move(pair.first)})};
+    dss_[1].rows = std::vector<Row>{Row(std::vector<Value>{std::move(pair.second)})};
+
+    resetPairState();
+    step_ = 0;
+    do {
+      // Set left(dss_[0]) or right(dss_[1]) dataset as req dataset, and get neighbors
+      for (int i = 0; i < 2; i++) {
+        if (break_ || step_ > shortestPathNode_->stepRange()->max()) {
+          dss_[0].clear();
+          dss_[1].clear();
+          break;
+        }
+        status = getNeighbors().get();
+        if (!status.ok()) {
+          return error(std::move(status));
+        }
+        setInterimState(1 - i);
+      }
+    } while (!dss_.empty() && !dss_[direction_].rows.empty());
+  }
+
+  return folly::makeFuture<Status>(buildResult());
+}
+
+void ShortestPathExecutor::setInterimState(int direction) {
+  direction_ = direction;
+  step_ += 1;
+}
+
+void ShortestPathExecutor::resetPairState() {
+  step_ = 0;
+  allPrevPaths_[0].clear();
+  allPrevPaths_[1].clear();
+  vidToVertex_.clear();
+  visiteds_[0].clear();
+  visiteds_[1].clear();
+  break_ = false;
+}
+
+Status ShortestPathExecutor::close() {
+  // clear the members
+  dss_[0].clear();
+  dss_[1].clear();
+  visiteds_[0].clear();
+  visiteds_[1].clear();
+  vidToVertex_.clear();
+  allPrevPaths_[0].clear();
+  allPrevPaths_[1].clear();
+
+  return Executor::close();
+}
+
+Status ShortestPathExecutor::buildRequestDataSet() {
+  auto inputVar = shortestPathNode_->inputVar();
+  const Result& inputResult = ectx_->getResult(inputVar);
```

Comment:
You can do the Cartesian product directly without the Cartesian product operator